### PR TITLE
Reduce comparisons count in merge sort

### PR DESCRIPTION
**Pull request by senior-zero**
In order to fix [the issue](https://github.com/NVIDIA/thrust/issues/1436) I've reduced the comparisons count in case of partially filled tile. The issue is caused by the fact that each thread has a fixed amount of data to process. If the input data size isn't a multiple of `ITEMS_PER_TILE`, there isn't enough data to fill the local storage of each thread. In this case, extra elements of the storage [are filled with some fixed value](https://github.com/senior-zero/thrust/blob/503806a36518fabf0ace0454deee903aee675614/thrust/system/cuda/detail/sort.h#L491-L494). After that the storage is sorted by `stable_odd_even_sort` by each thread (even ones outside the input range). The idea behind the fix is in skipping `stable_odd_even_sort` for threads of the last tile with no actual input data because in this case their storages are filled with the fixed value. 

The fix doesn't affect performance even for small input data sizes and complex comparators. For input data size of 2 INT32 elements number of comparisons is reduced by 507 times (RTX3090).

### CODE REVIEW

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

**Comment by allisonvacanti** [comment]
add to allowlist

**Comment by senior-zero** [comment]
run tests

File: thrust/system/cuda/detail/sort.h [hunk 1]
```diff
@@ -343,9 +344,10 @@ namespace __merge_sort {
         using core::uninitialized_array;
         using core::sync_threadblock;
 
-        // stable sort items in a single thread
+        // if first element of thread is in input range, stable sort items
         //
-        stable_odd_even_sort(keys_loc,items_loc);
+        if (!IS_LAST_TILE || ITEMS_PER_THREAD * tid < count)
```
Comment (allisonvacanti) [inline thrust/system/cuda/detail/sort.h:349]: Add braces

**Review by allisonvacanti — APPROVED** [review]

**Comment by allisonvacanti** [comment]
DVS CL: 30008797

run tests

**Pull request merged**